res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,718 @@
+        engine.setValue(group, 'loop_end_position', loopPoint.end);
+        engine.setValue(group, 'loop_in_goto', 1);
+        engine.setValue(group, 'reloop_toggle', 1);
+        // This is a hack, TODO find a command like loop_in_goto_and_play
+        if(engine.getValue(group, 'loop_enabled') == 0){
+            engine.setValue(group, 'reloop_toggle', 1);
+        }
+        return;
+    }
+
+    // save playing loop on hotcue pad
+    if(value > 0 && loopPlaying > 0){
+        const loopStart = engine.getValue(group, 'loop_start_position');
+        const loopEnd = engine.getValue(group, 'loop_end_position'); 
+        
+        this.hotcueLoopPoints[group][padNum-1] = {start: loopStart, end: loopEnd};
+        midi.sendShortMsg(status, control, 0x7f); // enable pad LED
+        return;
+    }
+
+};
+
+PioneerDDJ400.hotcuePadShiftPressed = function(_channel, control, value, status, group){
+    const loopPlaying = engine.getValue(group, 'loop_enabled');
+    const padNum = (control & 0xf) +1;
+    const loopPoint = this.hotcueLoopPoints[group][padNum-1];
+    
+    // shift is pressed -> delete hotcue or loop point
+    if(value > 0){
+        midi.sendShortMsg(status-1, control, 0); // disable Pad LED
+        engine.setValue(group, 'hotcue_'+padNum+'_clear', 1);
+        if ( loopPoint && loopPoint.start >=0 ){
+            this.hotcueLoopPoints[group][padNum-1] = {start: -1, end: -1};
+            if(loopPlaying > 0 ){
+                engine.setValue(group, 'reloop_toggle', 1);
+            }
+        }
+        return;
+    }
+};
+
+PioneerDDJ400.waveFormRotate = function(_channel, _control, value){
+    // select the Waveform to zoom left shift = deck1, right shift = deck2
+    const deckNum = this.shiftState[0] > 0 ? 1 : 2; 
+    const oldVal = engine.getValue('[Channel'+deckNum+']', 'waveform_zoom');
+    const newVal = oldVal + (value > 0x64 ? 1 : -1);
+    
+    engine.setValue('[Channel'+deckNum+']', 'waveform_zoom', newVal);
+};
+
+PioneerDDJ400.loopin4beatPressed = function(channel, _control, value, _status, group){
+    const loopEnabled = engine.getValue(group, 'loop_enabled');
+    this.loopin4beat[channel] = (value > 0);
+    
+    if(loopEnabled == 0 && value > 0){
+        engine.setValue(group, 'loop_in', 1);
+    }
+};
+
+PioneerDDJ400.loopin4beatPressedLong = function(_channel, _control, value, _status, group){
+    const loopEnabled = engine.getValue(group, 'loop_enabled');
+    if(!loopEnabled && value > 0){
+        engine.setValue(group, 'beatloop_4_activate', 1);
+    }
+};
+
+PioneerDDJ400.loopoutPressed = function(channel, _control, value, _status, group){
+    const loopEnabled = engine.getValue(group, 'loop_enabled');
+    this.loopout[channel] = (value > 0);
+
+    if(loopEnabled == 0 && value > 0){
+        engine.setValue(group, 'loop_out', 1);
+    }
+};
+
+// START BEAT FX
+
+PioneerDDJ400.numFxSlots = 3;
+
+Object.defineProperty(PioneerDDJ400, 'selectedFxSlot', {
+    get: function() {
```

Comment:
This works with the ancient QtScript interpreter? I remember trying this years ago and couldn't get it to work.